akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/ReceptionistMessages.scala

```diff
@@ -40,6 +46,19 @@ private[akka] object ReceptionistMessages {
       serviceInstance(key)
   }
 
+  final case class Unregistered[T] private[akka] (key: ServiceKey[T], _serviceInstance: ActorRef[T])
+      extends Receptionist.Unregistered {
+    def isForKey(key: ServiceKey[_]): Boolean = key == this.key
+    def serviceInstance[M](key: ServiceKey[M]): ActorRef[M] = {
+      if (key != this.key)
+        throw new IllegalArgumentException(s"Wrong key [$key] used, must use listing key [${this.key}]")
```

Comment:
Is the preference to do this with an `if` eval versus `require` to raise the `IllegalArgumentException`?

Reply:
Yes, to provide some context vs "requirement failed" or allocating a function for the lazy param.